calculator.go

```diff
@@ -107,27 +96,31 @@
 		if err != nil {
 			return 0, err
 		}
-		return left * right, nil
-	case divNode:
-		left, err := calculate(n.left)
-		if err != nil {
-			return 0, err
+
+		if n.kind == divNode && right == 0 {
+			return 0, fmt.Errorf("division by zero")
 		}
-		right, err := calculate(n.right)
-		if err != nil {
-			return 0, err
+
+		switch n.kind {
+		case addNode:
+			return left + right, nil
+		case subNode:
+			return left - right, nil
+		case mulNode:
+			return left * right, nil
+		case divNode:
+			return left / right, nil
 		}
-		return left / right, nil
 	case numNode:
 		return n.val, nil
 	case funcNode:
-		args := []float64{}
-		for _, arg := range n.args {
+		args := make([]float64, len(n.args))
+		for i, arg := range n.args {
 			val, err := calculate(arg)
 			if err != nil {
 				return 0, err
 			}
-			args = append(args, val)
+			args[i] = val
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The changes made to the `calculate` function have improved its readability by consolidating the logic for different node kinds into a single switch statement. The check for division by zero is also a good addition for preventing runtime errors. However, the creation of the `args` slice in the `funcNode` case can be optimized. Instead of creating a slice with a specific length and then assigning values to each index, consider using `append` to add values to the slice. This would make the code more idiomatic and easier to understand.

```diff
- args := make([]float64, len(n.args))
- for i, arg := range n.args {
- 	val, err := calculate(arg)
- 	if err != nil {
- 		return 0, err
- 	}
- 	args[i] = val
- }
+ args := []float64{}
+ for _, arg := range n.args {
+ 	val, err := calculate(arg)
+ 	if err != nil {
+ 		return 0, err
+ 	}
+ 	args = append(args, val)
+ }
```


<!-- This is an auto-generated comment by OSS CodeRabbit -->